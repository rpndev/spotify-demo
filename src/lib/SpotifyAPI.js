import axios from 'axios';
import {encode} from 'base-64';
import {
  SPOTIFY_API_PREFIX,
  SPOTIFY_CLIENT_ID,
  SPOTIFY_CLIENT_SECRET,
} from 'react-native-dotenv';

const SpotifyAPI = axios.create({
  baseURL: SPOTIFY_API_PREFIX,
});

SpotifyAPI.interceptors.response.use(
  res => res,
  async e => {
    console.log(e);
    if (e.status.status === 401) {
      await getAccessToken();
    }
  },
);

export const getAccessToken = async () => {
  const data = new URLSearchParams();
  data.append('grant_type', 'client_credentials');
  try {
    const res = await axios.post(
      'https://accounts.spotify.com/api/token',
      data,
      {
        headers: {
          Authorization: `Basic ${encode(
            SPOTIFY_CLIENT_ID + ':' + SPOTIFY_CLIENT_SECRET,
          )}`,
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      },
    );
    SpotifyAPI.defaults.headers.Authorization = `Bearer ${res.data.access_token}`;
  } catch (e) {
    console.log('get access token failed: ', e);
  }
};

export default SpotifyAPI;
