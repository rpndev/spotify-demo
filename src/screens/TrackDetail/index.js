/**
 * @format
 * @flow
 */

import React, {useState, useEffect} from 'react';
import {View, Text, Image, ActivityIndicator, StyleSheet} from 'react-native';
import SpotifyAPI from '../../lib/SpotifyAPI';

const TrackDetail = props => {
  const [loading, setLoading] = useState(true);
  const [track, setTrack] = useState({});
  useEffect(() => {
    SpotifyAPI.get(`/tracks/${props.navigation.state.params.track_id}`)
      .then(res => {
        setTrack(res.data);
        props.navigation.setParams({title: res.data.name});
        setLoading(false);
      })
      .catch(e => {
        setLoading(false);
        console.log('error', e.response);
      });
  }, [props]);

  return (
    <View style={styles.container}>
      {loading ? (
        <ActivityIndicator />
      ) : (
        <>
          <Text style={styles.trackNameText}>{track.name}</Text>
          <Text style={styles.artistsText}>
            Artists: {track.artists.map(e => e.name).join(', ')}
          </Text>
          <Text>Duration: {track.duration_ms}</Text>
          <View style={styles.albumContainer}>
            <Text>Album</Text>
            <Text>Name: {track.album.name}</Text>
            <Text>Total Tracks: {track.album.total_tracks}</Text>
            <Image
              source={{uri: track.album.images[0].url}}
              resizeMode="cover"
              style={styles.albumImage}
            />
          </View>
        </>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'pink',
    padding: 20,
  },
  trackNameText: {
    fontWeight: 'bold',
    color: 'red',
    fontSize: 24,
  },
  artistsText: {
    fontWeight: '400',
    fontSize: 20,
  },
  albumContainer: {
    marginHorizontal: 40,
    marginTop: 20,
    backgroundColor: 'white',
    borderRadius: 8,
    overflow: 'hidden',
  },
  albumImage: {
    height: 200,
    opacity: 0.7,
    padding: 12,
  },
});

TrackDetail.navigationOptions = ({navigation}) => ({
  title: navigation.state.params
    ? navigation.state.params.title
    : 'Track Detail',
});

export default TrackDetail;
