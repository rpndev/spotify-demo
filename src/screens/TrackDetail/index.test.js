import React from 'react';
import TrackDetail from './index';

import renderer from 'react-test-renderer';

import SpotifyAPI from '../../lib/SpotifyAPI';

function TrackData({ name, artists, duration_ms, tracks, image_url }) {
  return {
    name,
    artists,
    duration_ms,
    album: {
      total_tracks: tracks,
      images: [
        { url: image_url }
      ]
    }
  };
}

const stubbedAPIResponse = {
  name: 'Daytrip',
  artists: [
    { name: 'STEPHAN' },
    { name: 'Current Blue' }
  ],
  duration_ms: 180254,
  album: {
    total_tracks: 2,
    images: [
      { url: 'https://i.scdn.co/image/ac89987a987b98c978f978' }
    ]
  }
};

beforeEach(() => {
  SpotifyAPI.get = jest.fn(() => Promise.resolve({ data: { data: stubbedAPIResponse }}));
});

describe('TrackDetail', () => {
    it('renders correctly after fetching API', () => {
      const props = { 
        navigation: {
          navigate: jest.fn(),
          state: {
            params: {
              track_id: '23432432',
              title: 'asdfadsf'
            }
          },
          setParams: jest.fn()
          
        }
      };
      const tree = renderer.create(<TrackDetail {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
});