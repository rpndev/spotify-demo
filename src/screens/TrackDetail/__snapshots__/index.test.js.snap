// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayList renders correctly after fetching API 1`] = `
<View
  style={
    Object {
      "backgroundColor": "pink",
      "flex": 1,
      "padding": 20,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`TrackDetail renders correctly after fetching API 1`] = `
<View
  style={
    Object {
      "backgroundColor": "pink",
      "flex": 1,
      "padding": 20,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;
